### ISSUE TITLE
The problem of different broadening coefficients of multimolecules

### ISSUE THREAD

**Issue by Yuhao-9509** [radis/radis#740]
**Thank you very much for your hard work in maintaining and developing this project.** I really appreciate your efforts!

As mentioned in factory.py:
> before lineshape broadening (as it would not be constant for all species)"
it seems reasonable to first calculate each molecular species separately before applying broadening.

If computational time is not a major concern, would the following approach be a valid way to achieve a more accurate forward model spectrum?

```python
from radis import calc_spectrum, config
import numpy as np

config["MISSING_BROAD_COEF"] = "air"

s_N2O = calc_spectrum(
    wmin=2100, wmax=2350, 
    mole_fraction={"N2O": 1}, isotope="1,2,3",
    pressure=1.01325, Tgas=296, path_length=2400, 
    verbose=0, databank="hitran", mode='cpu', diluent="air"
)
nu_N2O, abscoef_N2O = np.array(s_N2O.get("abscoeff"))

s_H2O = calc_spectrum(
    wmin=2100, wmax=2350, 
    mole_fraction={"H2O": 1}, isotope="1,2,3",
    pressure=1.01325, Tgas=296, path_length=1, 
    verbose=0, databank="hitran", mode='cpu'
)
nu_H2O, abscoef_H2O = np.array(s_H2O.get("abscoeff"))

s_CO = calc_spectrum(
    wmin=2100, wmax=2350, 
    mole_fraction={"CO": 1}, isotope="1,2,3",
    pressure=1.01325, Tgas=296, path_length=1, 
    verbose=0, databank="hitran", mode='cpu'
)
nu_CO, abscoef_CO = np.array(s_CO.get("abscoeff"))

s_CO2 = calc_spectrum(
    wmin=2100, wmax=2350, 
    mole_fraction={"CO2": 1}, isotope="1,2,3",
    pressure=1.01325, Tgas=296, path_length=1, 
    verbose=0, databank="hitran", mode='cpu'
)
nu_CO2, abscoef_CO2 = np.array(s_CO2.get("abscoeff"))

# Compute total absorption coefficient
abscoef = (abscoef_N2O * 300e-9 + abscoef_H2O * 0.02 + 
           abscoef_CO * 100e-9 + abscoef_CO2 * 400e-6) * 2400

# Compute transmittance
transmittance_no = np.exp(-abscoef)

**Comment by minouHub** [comment]
Hello,
Indeed! The broadening is the main bottleneck for the moment. So, as suggested in the citation, it would be more efficient to sum the non-broadened spectra before applying the convolution. Not sure this is what your code is performing though. Can you elaborate?

**Comment by Yuhao-9509** [comment]
Thank you very much for your reply! I will briefly outline my ideas. Currently, my focus is solely on the aliasing of multiple gas components in the atmosphere, which is directly related to my research. In this case, I only need to consider atmospheric broadening, without involving more complex effects such as self-broadening.

Based on this premise, I am analyzing the atmospheric infrared spectra of:

- **400 ppm CO₂**  
- **2% H₂O**  
- **100 ppb CO**  
- **300 ppb N₂O**  

In this context, I noticed that the sample code first superimposes the temperature-corrected absorption intensities and then convolves them using the same Voigt line function. However, since **CO₂**, **CO**, **N₂O**, and **H₂O** have different molecular masses, their Voigt half-widths also differ. This results in discrepancies in the high-resolution absorbance or transmittance spectra generated by the two code implementations. 
The Doppler line shape function (Thermal motion of atmospheric molecules) is defined as:
$$
g_D\left(\nu\right) = \frac{1}{\alpha_D \sqrt{\pi}} \exp\left( -\frac{\left(\nu - \nu_{ij}\right)^2}{\alpha_D^2} \right)
$$

where the Doppler half-width at half-maximum (HWHM) is given by:

$$
\alpha_D = \nu_{ij} \sqrt{\frac{2 k_B T \ln{2}}{m c^2}}
$$
Additionally, I will cite your work in my article, including the relevant publication, as my forward model is based on your code. Thank you for your valuable contribution.

Below, I will provide my complete code for comparison.

'''python
from radis import calc_spectrum, config
import matplotlib.pyplot as plt
import numpy as np

config["MISSING_BROAD_COEF"] = "air"

s_N2O = calc_spectrum(
    wmin=2100,
    wmax=2350,  # cm-1
    # mole_fraction={"CO2": 400e-6, "H2O": 0.02, 'CO':100e-9, 'N2O':300e-9},
    mole_fraction={"N2O": 1},
    isotope="1,2,3",
    pressure=1.01325,  # bar
    Tgas=296,  # K
    path_length=2400,  # 1 km in cm
    verbose=0,
    databank="hitran",
    mode = 'cpu',
    diluent="air"
)
nu_N2O = np.array(s_N2O.get("abscoeff")[0])
abscoef_N2O = np.array(s_N2O.get("abscoeff")[1])

s_H2O = calc_spectrum(
    wmin=2100,
    wmax=2350,  # cm-1
    # mole_fraction={"CO2": 400e-6, "H2O": 0.02, 'CO':100e-9, 'N2O':300e-9},
    mole_fraction={"H2O": 1},
    isotope="1,2,3",
    pressure=1.01325,  # bar
    Tgas=296,  # K
    path_length=1,  # 1 km in cm
    verbose=0,
    databank="hitran",
    mode = 'cpu'
)

nu_H2O = np.array(s_H2O.get("abscoeff")[0])
abscoef_H2O = np.array(s_H2O.get("abscoeff")[1])

s_CO = calc_spectrum(
    wmin=2100,
    wmax=2350,  # cm-1
    # mole_fraction={"CO2": 400e-6, "H2O": 0.02, 'CO':100e-9, 'N2O':300e-9},
    mole_fraction={"CO": 1},
    isotope="1,2,3",
    pressure = 1.01325,  # bar
    Tgas = 296,  # K
    path_length = 1,  # 1 km in cm
    verbose = 0,
    databank="hitran",
    mode = 'cpu'
)
nu_CO = np.array(s_CO.get("abscoeff")[0])
abscoef_CO = np.array(s_CO.get("abscoeff")[1])

s_CO2 = calc_spectrum(
    wmin=2100,
    wmax=2350,  # cm-1
    # mole_fraction={"CO2": 400e-6, "H2O": 0.02, 'CO':100e-9, 'N2O':300e-9},
    mole_fraction = {"CO2": 1},
    isotope = "1,2,3",
    pressure = 1.01325,  # bar
    Tgas = 296,  # K
    path_length = 1,  # 1 km in cm
    verbose=0,
    databank="hitran",
    mode = 'cpu'
)

nu_CO2 = np.array(s_CO2.get("abscoeff")[0])
abscoef_CO2 = np.array(s_CO2.get("abscoeff")[1])

abscoef = (abscoef_N2O * 300e-9  + abscoef_H2O * 0.02 + abscoef_CO * 100e-9 + abscoef_CO2 * 400e-6) * 2400
# absorbance = abscoeff * new_path_length_cm  # calculate
transmittance_no = np.exp(-abscoef)

import time
s_time =time.time()
s = calc_spectrum(
    wmin=2100,
    wmax=2350,  # cm-1
    mole_fraction={"CO2": 400e-6, "H2O": 0.02, 'CO':100e-9, 'N2O':300e-9},
    # mole_fraction={"N2O": 300e-9},
    isotope="1,2,3",
    pressure=1.01325,  # bar
    Tgas=296,  # K
    path_length=2400,  # 1 km in cm
    verbose=0,
    databank="hitran",
    mode = 'cpu'
)
e_time = time.time()
print(e_time - s_time)
nu = np.array(s.get("transmittance_noslit")[0])
#print(nu)
transmittance_noslit = np.array(s.get("transmittance_noslit")[1])
plt.plot(nu, transmittance_noslit)
plt.plot(nu, transmittance_no)
plt.show()

s.apply_slit(1, "cm-1")
# nu = np.array(s.get("transmittance")[0])
# transmittance = np.array(s.get("transmittance")[1])
# plt.plot(nu, transmittance)
# plt.show()


# print(s.get_slit(wunit="same")[1])


import numpy as np
import matplotlib.pyplot as plt

def triangular_slit(FWHM, wstep, center = 0):
    """生成三角形狭缝函数"""
    slope = 1 / (FWHM - wstep)
    a = int(FWHM / wstep)
    I = 1 - np.arange(0, a) * slope * wstep  # 斜率

    # 添加零点
    I = np.hstack((I[::-1], I[1:]))
    I /= np.sum(I)
    # w = np.linspace(-a * wstep, a * wstep, len(I)) + center

    return I

# 示例
FWHM = 1  # 全宽半高
wstep = 0.01  # 波长步长
center = 0  # 中心波长
I = triangular_slit(FWHM, wstep, center)
print(type(I))

# 绘制狭缝函数
plt.plot(I, label='Triangular Slit Function')
plt.xlabel('Wavelength (nm)')
plt.ylabel('Intensity')
plt.legend()
plt.show()
print(I)

# print(np.allclose(np.array(s.get_slit(wunit="same")[1]), I))

result = np.convolve(transmittance_no, I, mode='valid')
print(transmittance_no)
print(triangular_slit)
plt.plot(result)
plt.plot(np.array(s.get("transmittance")[1]))
plt.show()

**Comment by Yuhao-9509** [comment]
> Hello, Indeed! The broadening is the main bottleneck for the moment. So, as suggested in the citation, it would be more efficient to sum the non-broadened spectra before applying the convolution. Not sure this is what your code is performing though. Can you elaborate?


### Atmospheric Infrared Spectra Analysis

Based on this premise, I am analyzing the atmospheric infrared spectra of:

- **400 ppm CO₂**  
- **2% H₂O**  
- **100 ppb CO**  
- **300 ppb N₂O**  

In this context, I noticed that the sample code first superimposes the temperature-corrected absorption intensities and then convolves them using the same Voigt line function. However, since **CO₂**, **CO**, **N₂O**, and **H₂O** have different molecular masses, their Voigt half-widths also differ. This results in discrepancies in the high-resolution absorbance or transmittance spectra generated by the two code implementations.

### Doppler Line Shape Function

The Doppler line shape function, accounting for the thermal motion of atmospheric molecules, is defined as:

$$
g_D\\left(\\nu\\right) = \\frac{1}{\\alpha_D \\sqrt{\\pi}} \\exp\\left( -\\frac{\\left(\\nu - \\nu_{ij}\\right)^2}{\\alpha_D^2} \\right)
$$

Where the Doppler half-width at half-maximum (HWHM) is:

$$
\\alpha_D = \\nu_{ij} \\sqrt{\\frac{2 k_B T \\ln{2}}{m c^2}}
$$

### Code Implementation

Additionally, I will cite your work in my article, including the relevant publication, as my forward model is based on your code. Thank you for your valuable contribution.

Below is the complete code for comparison:

```python
from radis import calc_spectrum, config
import matplotlib.pyplot as plt
import numpy as np

# Configuration
config["MISSING_BROAD_COEF"] = "air"

# Function to calculate spectrum for different molecules
def calculate_spectrum(molecule, mole_fraction, path_length):
    return calc_spectrum(
        wmin=2100,
        wmax=2350,  # cm-1
        mole_fraction={molecule: mole_fraction},
        isotope="1,2,3",
        pressure=1.01325,  # bar
        Tgas=296,  # K
        path_length=path_length,  # path length in cm
        verbose=0,
        databank="hitran",
        mode='cpu'
    )

# Calculate absorption coefficients for each molecule
s_N2O = calculate_spectrum("N2O", 1, 2400)
s_H2O = calculate_spectrum("H2O", 1, 1)
s_CO = calculate_spectrum("CO", 1, 1)
s_CO2 = calculate_spectrum("CO2", 1, 1)

# Extract absorption coefficients
nu_N2O, abscoef_N2O = np.array(s_N2O.get("abscoeff")[0]), np.array(s_N2O.get("abscoeff")[1])
nu_H2O, abscoef_H2O = np.array(s_H2O.get("abscoeff")[0]), np.array(s_H2O.get("abscoeff")[1])
nu_CO, abscoef_CO = np.array(s_CO.get("abscoeff")[0]), np.array(s_CO.get("abscoeff")[1])
nu_CO2, abscoef_CO2 = np.array(s_CO2.get("abscoeff")[0]), np.array(s_CO2.get("abscoeff")[1])

# Combine absorption coefficients
abscoef = (abscoef_N2O * 300e-9 + abscoef_H2O * 0.02 + abscoef_CO * 100e-9 + abscoef_CO2 * 400e-6) * 2400
transmittance_no = np.exp(-abscoef)  # Calculate transmittance

# Direct spectrum calculation for comparison
import time
s_time = time.time()
s = calc_spectrum(
    wmin=2100,
    wmax=2350,  # cm-1
    mole_fraction={"CO2": 400e-6, "H2O": 0.02, "CO": 100e-9, "N2O": 300e-9},
    isotope="1,2,3",
    pressure=1.01325,  # bar
    Tgas=296,  # K
    path_length=2400,  # cm
    verbose=0,
    databank="hitran",
    mode='cpu'
)
e_time = time.time()
print(f"Calculation time: {e_time - s_time} seconds")

# Plotting the spectra
nu = np.array(s.get("transmittance_noslit")[0])
transmittance_noslit = np.array(s.get("transmittance_noslit")[1])

plt.plot(nu, transmittance_noslit, label='Direct Calculation')
plt.plot(nu, transmittance_no, label='Superimposed Spectra')
plt.xlabel('Wavenumber (cm$^{-1}$)')
plt.ylabel('Transmittance')
plt.legend()
plt.show()

# Applying a triangular slit function
def triangular_slit(FWHM, wstep):
    slope = 1 / (FWHM - wstep)
    a = int(FWHM / wstep)
    I = 1 - np.arange(0, a) * slope * wstep  
    I = np.hstack((I[::-1], I[1:]))
    I /= np.sum(I)  # Normalize
    return I

# Generate slit function
FWHM = 1  # cm-1
wstep = 0.01  # cm-1
I = triangular_slit(FWHM, wstep)
print(f"Slit function generated with type: {type(I)}")
```

**Comment by minouHub** [comment]
Alright so it seems that the problems comes from the Doppler broadening (Gaussian). The Lorentzian should not be affected by the molecular masses.

**Comment by erwanp** [comment]
hello @Yuhao-9509 ,see this discussion https://github.com/radis/radis/767 about the role of neighbour_lines. Setting it to the same value as the truncation is a consistent choice.

also note that larger truncation does not necessarily mean more accurate results, real lineshapes far wings have a sub-lorentzian behavior which is better reproduced by a truncated voigt/lorentzian lineshape than a non truncated one. We chose 50 cm/1 as a good trade off. Researcher looking for extreme lineshape accuracy typically use chi factors (easy to implement see the code here, but it was never pushed https://github.com/radis/radis/pull/533), or non-Voigt lineshapes. For the majority of the use cases this is not needed, I just share it for the sake of completeness.

**Comment by minouHub** [comment]
> Would this be appropriate if the instrument line shape function is not taken into account?

Not sure you can do high-accuracy measurements without the instrumental broadening.

> Additionally, if we do consider the instrument line shape function, should we take into account the convolution of the sinc function with the N-B apodization?

You need to apply by hand the sinc broadening, see #284. The "the N-B apodization" is beyond my expertise.

**Comment by Yuhao-9509** [comment]
> > Would this be appropriate if the instrument line shape function is not taken into account?
> 
> Not sure you can do high-accuracy measurements without the instrumental broadening.
> 
> > Additionally, if we do consider the instrument line shape function, should we take into account the convolution of the sinc function with the N-B apodization?
> 
> You need to apply by hand the sinc broadening, see [#284](https://github.com/radis/radis/issues/284). The "the N-B apodization" is beyond my expertise.

Really appreciate your help—and thanks again for your awesome work on this open source project!

**Comment by Yuhao-9509** [comment]
Since I use a 1 cm⁻¹ FTIR spectrometer, I typically don't use Gaussian functions as the instrument line shape (ILS), but rather sinc functions, which are the Fourier transforms of rectangular functions. 

For implementation, we referred to **HAPI** and used **PyTorch** for a basic version. I also discovered that the **Norton–Beer (N-B) apodization function** provides a corresponding Python interface, which can be used to implement this functionality. Extension of the Norton–Beer apodizing functions in Fourier transform spectrometry:https://opg.optica.org/ao/abstract.cfm?uri=ao-51-20-4622

While this may be somewhat tangential to the main topic of **Radis**, it could serve as a useful reference for others who encounter similar questions and might help facilitate further discussion. The relevant code is as follows:

```python
def SLIT_MICHELSON_torch(step=0.007531, FWHM=1.5, device='cuda:0'):
    start, stop = -20, 20
    num = int(round((stop - start) / step)) + 1
    x = torch.linspace(-20, 20, num, device=device, dtype=torch.float64)
    y = torch.zeros_like(x)
    dk_ = 2 * torch.pi / FWHM
    x_ = dk_ * x
    index_zero = x == 0
    index_nonzero = torch.logical_not(index_zero)
    y[index_zero] = 2 / FWHM
    y[index_nonzero] = (2 / FWHM) * torch.sin(x_[index_nonzero]) / x_[index_nonzero]
    y /= y.sum()
